### PR TITLE
Support passing `storage_options` in `trainer.save_checkpoint()` API

### LINKED ISSUE #10629: Support passing `storage_options` in `trainer.save_checkpoint()` API
### Motivation

Support specifying `storage_options` parameter in `trainer.save_checkpoint()` API.
For metadata that varies across different checkpoint instances, users can then specify `trainer.save_checkpoint(storage_options=...)` per individual checkpoint instance saved, that ultimately gets passed to CheckpointIOPlugin to inform how to best save that particular checkpoint instance.

### Pitch

Introduce new optional `storage_options` arg in `trainer.save_checkpoint()`:
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L1864-L1865

Introduce and pass arg to `checkpoint_connector.save_checkpoint()`:
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/checkpoint_connector.py#L453

Introduce and pass arg to `TTP.save_checkpoint()`:
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/training_type_plugin.py#L294

Pass arg (arg already exists) to `CheckpointIOPlugin.save_checkpoint()`:
https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/io/checkpoint_plugin.py#L36

### Alternatives

Having to specify through `trainer.training_type_plugin.checkpoint_io.save_checkpoint(storage_options=...)`

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @awaelchli @ananthsub @ninginthecloud @rohitgr7

### PR DESCRIPTION
**Pull request by jjenniferdai**
## What does this PR do?

Fixes #10629

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/strategies/deepspeed.py [hunk 1]
```diff
@@ -728,12 +728,13 @@ def deepspeed_engine(self):
...
         return self.num_processes > 1 or self.num_nodes > 1
 
-    def save_checkpoint(self, checkpoint: Dict, filepath: _PATH) -> None:
+    def save_checkpoint(self, checkpoint: Dict, filepath: _PATH, storage_options: Optional[Any] = None) -> None:
         """Save model/training states as a checkpoint file through state-dump and file-write.
 
         Args:
             checkpoint: The checkpoint state dictionary
             filepath: write-target file's path
+            storage_options: not used for ``DeepSpeedStrategy`` as ``CheckpointIO`` is not used
```
Comment (ananthsub) [inline pytorch_lightning/strategies/deepspeed.py:737]: should this raise an exception if storage options are passed?

File: pytorch_lightning/strategies/strategy.py [hunk 2]
```diff
@@ -402,15 +402,18 @@ def lightning_module_state_dict(self) -> Dict[str, Union[Any, Tensor]]:
...
-    def save_checkpoint(self, checkpoint: Dict[str, Any], filepath: _PATH) -> None:
+    def save_checkpoint(
+        self, checkpoint: Dict[str, Any], filepath: _PATH, storage_options: Optional[Any] = None
+    ) -> None:
         """Save model/training states as a checkpoint file through state-dump and file-write.
 
         Args:
             checkpoint: dict containing model and trainer state
             filepath: write-target file's path
+            storage_options: Optional parameter for how to save to storage, passed to ``CheckpointIO`` plugin
```
Comment (ananthsub) [inline pytorch_lightning/strategies/strategy.py:413]: "Optional" is redundant since that's included in the typehint

File: pytorch_lightning/strategies/strategy.py [hunk 3]
```diff
@@ -402,15 +402,18 @@ def lightning_module_state_dict(self) -> Dict[str, Union[Any, Tensor]]:
...
         """Save model/training states as a checkpoint file through state-dump and file-write.
 
         Args:
             checkpoint: dict containing model and trainer state
             filepath: write-target file's path
+            storage_options: parameter for how to save to storage, passed to ``CheckpointIO`` plugin
         """
         if self.is_global_zero:
-            self.checkpoint_io.save_checkpoint(checkpoint, filepath)
+            self.checkpoint_io.save_checkpoint(checkpoint, filepath, storage_options)
```
Comment (carmocca) [inline pytorch_lightning/strategies/strategy.py:416]: I just noticed `checkpoint_io.save_checkpoint` takes the argument but does not use it 😕

File: pytorch_lightning/strategies/strategy.py [hunk 4]
```diff
@@ -402,15 +402,18 @@ def lightning_module_state_dict(self) -> Dict[str, Union[Any, Tensor]]:
         model = self.lightning_module
         return model.state_dict()
 
-    def save_checkpoint(self, checkpoint: Dict[str, Any], filepath: _PATH) -> None:
+    def save_checkpoint(
```
Comment (carmocca) [inline pytorch_lightning/strategies/strategy.py:405]: A simple mock test for the base strategy would be appreciated

`mock_ckpt_io.assert_called_once_with(...)`

File: pytorch_lightning/strategies/deepspeed.py [hunk 5]
```diff
@@ -728,13 +728,20 @@ def deepspeed_engine(self):
...
 
-    def save_checkpoint(self, checkpoint: Dict, filepath: _PATH) -> None:
+    def save_checkpoint(self, checkpoint: Dict, filepath: _PATH, storage_options: Optional[Any] = None) -> None:
         """Save model/training states as a checkpoint file through state-dump and file-write.
 
         Args:
             checkpoint: The checkpoint state dictionary
             filepath: write-target file's path
+            storage_options: raises exception if passed in,
+                not used for ``DeepSpeedStrategy`` as ``CheckpointIO`` is not used
```
Comment (ananthsub) [inline pytorch_lightning/strategies/deepspeed.py:738]: See the docstring style here for desribing exceptions: https://github.com/PyTorchLightning/pytorch-lightning/blob/79c4e5de60685dbc895641b0139ffc6180d069aa/pytorch_lightning/utilities/finite_checks.py#L34-L43

File: pytorch_lightning/strategies/strategy.py (hunk 3 above)
Comment (jjenniferdai) [reply pytorch_lightning/strategies/strategy.py]: you mean `TorchCheckpointIO.save_checkpoint`?

File: pytorch_lightning/strategies/tpu_spawn.py [hunk 6]
```diff
@@ -304,15 +304,18 @@ def _pod_progress_bar_force_stdout(self) -> None:
...
         """Save model/training states as a checkpoint file through state-dump and file-write.
 
         Args:
             checkpoint: dict containing model and trainer state
             filepath: write-target file's path
+            storage_options: parameter for how to save to storage, passed to ``CheckpointIO`` plugin
         """
         # `xla_model.save` needs to be called on all ranks. It internally checks if the local rank is 0
-        self.checkpoint_io.save_checkpoint(checkpoint, filepath)
+        self.checkpoint_io.save_checkpoint(checkpoint, filepath, storage_options)
```
Comment (kaushikb11) [inline pytorch_lightning/strategies/tpu_spawn.py:318]: Couldn't support `storage_options` for tpu_spawn strategy. We use the xla API instead of `fsspec` for checkpoint saving.

Comment (kaushikb11) [reply pytorch_lightning/strategies/tpu_spawn.py:318]: But yeah, users should have the flexibility to subclass XLACheckpointIO plugin and use `storage_options`, the way they desire.

**Comment by kaushikb11** [comment]
@jjenniferdai Any updates on the PR?

**Comment by jjenniferdai** [comment]
re: `XLACheckpointIO` and `TorchCheckpointIO` not using `storage_options`: should we leave the implementations as is (and maybe add a warning if `storage_options` is not None)? 

`storage_options` can be used for users to pass to their own custom `CheckpointIO` --> own `CheckpointIO` defines exactly how to use it

File: pytorch_lightning/strategies/tpu_spawn.py (hunk 6 above)
Comment (jjenniferdai) [reply pytorch_lightning/strategies/tpu_spawn.py:265]: > users should have the flexibility to subclass `XLACheckpointIO ` and use `storage_options`, the way they desire.

agree - in general, to subclass `CheckpointIO`

**Comment by kaushikb11** [comment]
> add a warning if storage_options is not None.

IMO, yes. We should add a warning for it for the users.

**Review by kaushikb11 — APPROVED** [review]

File: pytorch_lightning/plugins/io/torch_plugin.py [hunk 7]
```diff
@@ -30,6 +31,23 @@ class TorchCheckpointIO(CheckpointIO):
...
+            path: write-target path
+            storage_options: not used in ``TorchCheckpointIO.save_checkpoint``
+
+        Raises:
+            MisconfigurationException:
+                If ``storage_options`` arg is passed in
+        """
+        if storage_options is not None:
+            raise MisconfigurationException(
+                "`Trainer.save_checkpoint(storage_options)` with `storage_options` arg "
```
Comment (carmocca) [inline pytorch_lightning/plugins/io/torch_plugin.py:47]: Same for the other classes

```suggestion
                "`Trainer.save_checkpoint(..., storage_options=...)` "
```

File: pytorch_lightning/plugins/io/torch_plugin.py [hunk 8]
```diff
@@ -30,6 +31,23 @@ class TorchCheckpointIO(CheckpointIO):
...
+                "is not supported for `TorchCheckpointIO`. Please implement your custom `CheckpointIO`"
```
Comment (carmocca) [inline pytorch_lightning/plugins/io/torch_plugin.py:48]: Same for the other classes

```suggestion
                "is not supported for `{self.__class__.__name__}`. Please implement your custom `CheckpointIO`"
```

File: pytorch_lightning/strategies/tpu_spawn.py [hunk 9]
```diff
@@ -251,15 +251,18 @@ def _pod_progress_bar_force_stdout(self) -> None:
...
```
Comment (carmocca) [inline pytorch_lightning/strategies/tpu_spawn.py:265]:
```suggestion
        self.checkpoint_io.save_checkpoint(checkpoint, filepath, storage_options=storage_options)
```

File: pytorch_lightning/trainer/trainer.py [hunk 10]
```diff
@@ -2332,16 +2332,20 @@ def predicted_ckpt_path(self, ckpt_path: Optional[str]) -> None:
...
+        self, filepath: _PATH, weights_only: bool = False, storage_options: Optional[Any] = None
+    ) -> None:
         r"""
         Runs routine to create a checkpoint.
 
         Args:
             filepath: Path where checkpoint is saved.
             weights_only: If ``True``, will only save the model weights.
+            storage_options: parameter for how to save to storage, passed to ``CheckpointIO`` plugin
+                Define in your custom ``CheckpointIO`` plugin how you'd like to use any passed in ``storage_options``.
```
Comment (carmocca) [inline pytorch_lightning/trainer/trainer.py:2345]: The othe docstrings do not include this description

```suggestion
```

**Review by carmocca — COMMENTED** [review]
@jjenniferdai, @ananthsub and I chose this for 1.7 so we had time to see a fuller picture for how to implement storage options across the codebase.

**Comment by jjenniferdai** [comment]
> fuller picture for how to implement storage options across the codebase.

e.g. like the other filesystem ops in `ModelCheckpoint`?
- this is the only entry point to enable different storage_options across different checkpoint instances right?
- I think this is also different than just fsspec "storage_options" - it's really just for passing any checkpoint-instance specific information for `CheckpointIO` to use however it'd like to - thoughts?

File: pytorch_lightning/plugins/io/xla_plugin.py [hunk 11]
```diff
@@ -36,8 +37,18 @@ def save_checkpoint(self, checkpoint: Dict[str, Any], path: _PATH, storage_optio
...
+
+        Raises:
+            MisconfigurationException:
+                If ``storage_options`` arg is passed in
         """
+        if storage_options is not None:
+            raise MisconfigurationException(
+                "`Trainer.save_checkpoint(..., storage_options=...)` with `storage_options` arg "
+                f"is not supported for `{self.__class__.__name__}`. Please implement your custom `CheckpointIO`"
+                "to define how you'd like to use `storage_options`."
```
Comment (awaelchli) [inline pytorch_lightning/plugins/io/xla_plugin.py:50]:
```suggestion
                " to define how you'd like to use `storage_options`."
```
space is missing, in other messages too.
best is to put it in front so it doesn't get missed.

make sure to trigger the error at least once to inspect it :)) don't trust `pytest.raises`! 😄

**Review by awaelchli — APPROVED** [review]

**Review by ananthsub — APPROVED** [review]

**Comment by kaushikb11** [comment]
@tchaton @Borda @SeanNaren Mind reviewing?

File: pytorch_lightning/plugins/io/torch_plugin.py [hunk 12]
```diff
@@ -30,6 +31,23 @@ class TorchCheckpointIO(CheckpointIO):
...
+            checkpoint: dict containing model and trainer state
+            path: write-target path
+            storage_options: not used in ``TorchCheckpointIO.save_checkpoint``
+
+        Raises:
+            MisconfigurationException:
+                If ``storage_options`` arg is passed in
+        """
+        if storage_options is not None:
+            raise MisconfigurationException(
```
Comment (carmocca) [inline pytorch_lightning/plugins/io/torch_plugin.py:46]: #10918 

```suggestion
            raise TypeError(
```

Same comment applies to the other usages

File: tests/checkpointing/test_trainer_checkpoint.py [hunk 13]
```diff
@@ -87,3 +93,49 @@ def test_accumulated_gradient_batches_with_ckpt_path(tmpdir):
...
+
+
+def test_trainer_save_checkpoint_storage_options(tmpdir):
+    """This test validates that storage_options argument is properly passed to ``CheckpointIO``"""
+    model = BoringModel()
+    trainer = Trainer(
+        default_root_dir=tmpdir,
+        max_epochs=1,
+        limit_train_batches=1,
+        limit_val_batches=0,
```
Comment (carmocca) [inline tests/checkpointing/test_trainer_checkpoint.py:105]: Would this work?

```suggestion
        fast_dev_run=True
```

**Review by carmocca — APPROVED** [review]

File: tests/checkpointing/test_trainer_checkpoint.py (hunk 13 above)
Comment (jjenniferdai) [reply tests/checkpointing/test_trainer_checkpoint.py]: ah yes! the fast_dev_run skip saving check is only in `ModelCheckpoint` callback, not the actual `trainer.save_checkpoint` call
